config/config.go

```diff
@@ -0,0 +1,237 @@
+			cfgUser.SetValue(mysqldUser)
+		}
+	}
+
+	cfg.ValueMapper = os.ExpandEnv
+	config := &Config{}
+	m := make(map[string]MySqlConfig)
+	for _, sec := range cfg.Sections() {
+		sectionName := sec.Name()
+
+		if sectionName == "DEFAULT" {
+			continue
+		}
+
+		mysqlcfg := &MySqlConfig{
+			TlsInsecureSkipVerify: tlsInsecureSkipVerify,
+		}
+		if err != nil {
+			level.Error(logger).Log("msg", "failed to load config", "section", sectionName, "err", err)
+			continue
+		}
+
+		err = sec.StrictMapTo(mysqlcfg)
+		if err != nil {
+			level.Error(logger).Log("msg", "failed to parse config", "section", sectionName, "err", err)
+			continue
+		}
+		if err := mysqlcfg.validateConfig(); err != nil {
+			level.Error(logger).Log("msg", "failed to validate config", "section", sectionName, "err", err)
+			continue
+		}
+
+		m[sectionName] = *mysqlcfg
+	}
+	config.Sections = m
+	if len(config.Sections) == 0 {
+		return fmt.Errorf("no configuration found")
+	}
+	ch.Lock()
+	ch.Config = config
+	ch.Unlock()
+	return nil
+}
+
+func (m MySqlConfig) validateConfig() error {
+	if m.User == "" {
+		return fmt.Errorf("no user specified in section or parent")
+	}
+	if m.Password == "" {
+		return fmt.Errorf("no password specified in section or parent")
+	}
+
+	return nil
+}
+
+func (m MySqlConfig) FormDSN(target string) (string, error) {
+	var dsn string
+
+	user := m.User
+	password := m.Password
+	if target == "" {
+		host := m.Host
+		port := m.Port
+		socket := m.Socket
+		if socket != "" {
+			dsn = fmt.Sprintf("%s:%s@unix(%s)/", user, password, socket)
+		} else {
+			dsn = fmt.Sprintf("%s:%s@tcp(%s:%d)/", user, password, host, port)
+		}
+	} else {
+		host, port := parseAddress(target)
+		dsn = fmt.Sprintf("%s:%s@tcp(%s:%s)/", user, password, host, port)
+	}
+
+	if m.SslCa != "" {
+		if tlsErr := m.CustomizeTLS(); tlsErr != nil {
+			tlsErr = fmt.Errorf("failed to register a custom TLS configuration for mysql dsn: %s", tlsErr)
+			return dsn, tlsErr
+		}
+		dsn = fmt.Sprintf("%s?tls=custom", dsn)
+	}
+
+	return dsn, nil
+}
+
+func (m MySqlConfig) CustomizeTLS() error {
+	var tlsCfg tls.Config
+	caBundle := x509.NewCertPool()
+	pemCA, err := os.ReadFile(m.SslCa)
+	if err != nil {
+		return err
+	}
+	if ok := caBundle.AppendCertsFromPEM(pemCA); ok {
+		tlsCfg.RootCAs = caBundle
+	} else {
+		return fmt.Errorf("failed parse pem-encoded CA certificates from %s", m.SslCa)
+	}
+	if m.SslCert != "" && m.SslKey != "" {
+		certPairs := make([]tls.Certificate, 0, 1)
+		keypair, err := tls.LoadX509KeyPair(m.SslCert, m.SslKey)
+		if err != nil {
+			return fmt.Errorf("failed to parse pem-encoded SSL cert %s or SSL key %s: %s",
+				m.SslCert, m.SslKey, err)
+		}
+		certPairs = append(certPairs, keypair)
+		tlsCfg.Certificates = certPairs
+	}
+	tlsCfg.InsecureSkipVerify = m.TlsInsecureSkipVerify
+	mysql.RegisterTLSConfig("custom", &tlsCfg)
+	return nil
+}
+
+func parseAddress(address string) (string, string) {
+	target := strings.Split(address, ":")
```

Comment:
I'd prefer to use Go `net.SplitHostPort()`. This will break with IPv6 addresses.

https://pkg.go.dev/net#SplitHostPort